Sources/TSCBasic/FileSystem.swift

```diff
@@ -535,9 +544,16 @@ public class InMemoryFileSystem: FileSystem {
         }
     }
 
-    /// The root filesytem.
+    /// The root node of the filesytem.
     private var root: Node
-    /// Used to access lockFiles in a thread safe manner.
+    
+    /// Protects `root` and everything underneath it.
+    /// FIXME: Using a single lock for this is a performance problem, but in
+    /// reality, the only practical use for InMemoryFileSystem is for unit
+    /// tests.
+    private let lock = Lock()
+    
+    /// Exclusive file system lock vended to clients through `withLock()`.
     private let lockFilesLock = Lock()
```

Comment:
I think we could use replace uses of `lockFilesLock ` with `lock`

Reply:
Actually, that was the first think I tried — but they really are locking different things.  One is the internal lock for protecting `root`, and the other is the external client lock.  If only one lock is used for this, we have deadlock, since it means that once you've locked the file system, you can't also do any file system operations (`Lock` isn't recursive).

Reply:
You can see this if you replace use of `lockFilesLock` with `lock` in `withLock()`.  This will cause a couple of the tests to deadlock.

Reply:
Also, I think with the overall file system lock now in place, I think we can probably restore the ability to do per-file locking as a follow-on PR.